tsdate/core.py

```diff
@@ -957,47 +939,104 @@ def __init__(
             if Ne is not None:
-                raise ValueError(
-                    "Cannot specify population size if specifying priors "
-                    f"from tsdate.build_{self.prior_grid_func_name}()"
+                raise ValueError(f"Population size is not used for method {self.name}")
+        else:
+            if priors is None:
+                if Ne is None:
+                    raise ValueError(
+                        "Must specify population size if priors are not already "
+                        f"built using tsdate.build_{self.prior_grid_func_name}()"
+                    )
+                mk_prior = getattr(prior, self.prior_grid_func_name)
+                # Default to not creating approximate priors unless ts has
+                # greater than DEFAULT_APPROX_PRIOR_SIZE samples
+                approx = (
+                    True if ts.num_samples > base.DEFAULT_APPROX_PRIOR_SIZE else False
+                )
+                self.priors = mk_prior(
+                    ts, Ne, approximate_priors=approx, progress=progress
                 )
-            self.priors = priors
+            else:
+                logging.info("Using user-specified priors")
+                if Ne is not None:
+                    raise ValueError(
+                        "Cannot specify population size if specifying priors "
+                        f"from tsdate.build_{self.prior_grid_func_name}()"
+                    )
+                self.priors = priors
+
+        # mutation to edge mapping
+        mutspan_timing = time.time()
+        self.edges_mutations, self.mutations_edge = util.mutation_span_array(ts)
+        mutspan_timing -= time.time()
+        logging.info(f"Extracted mutations in {abs(mutspan_timing)} seconds")
 
     def get_modified_ts(self, result, eps):
         # Return a new ts based on the existing one, but with the various
         # time-related information correctly set.
         tables = self.ts.dump_tables()
         nodes = tables.nodes
+        mutations = tables.mutations
         if self.provenance_params is not None:
             provenance.record_provenance(tables, self.name, **self.provenance_params)
-        nodes.time = constrain_ages_topo(self.ts, result.posterior_mean, eps, self.pbar)
+        # Constrain node ages for positive branch lengths
+        constr_timing = time.time()
+        nodes.time = util.constrain_ages(
+            self.ts, result.posterior_mean, eps, self.constr_iterations
+        )
+        mutations.time = util.constrain_mutations(
+            self.ts, nodes.time, self.mutations_edge
+        )
         tables.time_units = self.time_units
-        tables.mutations.time = np.full(self.ts.num_mutations, tskit.UNKNOWN_TIME)
-        # Add posterior mean and variance to node metadata
-        if result.posterior_obj is not None:
+        constr_timing -= time.time()
+        logging.info(f"Constrained node ages in {abs(constr_timing)} seconds")
```

Comment:
Maybe more useful to output an (optional) tqdm bar to give feedback, which presumably would remove the need to report times?

Reply:
The constraint is all done in numba so no tqdm bar. I'd prefer to keep logging things as I do here.

Reply:
Fair enough.